crates/flipperzero/src/bluetooth/beacon.rs

```diff
@@ -0,0 +1,661 @@
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            major_bytes[0],
+            major_bytes[1],
+            minor_bytes[0],
+            minor_bytes[1],
+            signal_power,
+        ];
+        mf_data[4..20].copy_from_slice(uuid.as_bytes());
+
+        Self::empty()
+            // LE and BR/EDR flag.
+            .with(AdType::Flags, [0x06])
+            .expect("within size limits")
+            // iBeacon AD structure
+            .with(AdType::ManufacturerSpecificData, mf_data)
+            .expect("within size limits")
+    }
+
+    /// Constructs an [AltBeacon] Advertisement Data packet.
+    ///
+    /// > For interoperability purposes, the first 16+ bytes of the beacon identifier
+    /// > should be unique to the advertiser's organizational unit. Any remaining bytes of
+    /// > the beacon identifier may be subdivided as needed for the use case.
+    ///
+    /// `ref_rssi` is the average RSSI at 1 metre.
+    ///
+    /// Returns `None` if `ref_rssi` is not between -127 and 0 inclusive.
+    ///
+    /// [AltBeacon]: https://github.com/AltBeacon/spec
+    pub fn alt_beacon(
+        mfg_id: u16,
+        beacon_id: [u8; 20],
+        ref_rssi: i8,
+        mfg_reserved: u8,
+    ) -> Option<Self> {
+        if !(-127..=0).contains(&ref_rssi) {
+            return None;
+        }
+
+        let mfg_id_bytes = mfg_id.to_le_bytes();
+        let mut mf_data = [
+            mfg_id_bytes[0],
+            mfg_id_bytes[1],
+            0xBE,
+            0xAC,
+            0x00, // Beacon ID
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            0x00,
+            ref_rssi as u8,
+            mfg_reserved,
+        ];
+        mf_data[4..24].copy_from_slice(&beacon_id);
+
+        Some(
+            Self::empty()
+                // AltBeacon AD structure
+                .with(AdType::ManufacturerSpecificData, mf_data)
+                .expect("within size limits"),
+        )
+    }
+
+    /// Constructs an [Eddystone-URL] Advertisement Data packet.
+    ///
+    /// [Eddystone-URL]: https://github.com/google/eddystone/tree/master/eddystone-url
+    #[cfg(feature = "alloc")]
+    pub fn eddystone_url(tx_power: u8, url_scheme: EddystoneUrlScheme, url: &str) -> Self {
+        let mut service_data = alloc::vec![
+            0xAA, // Eddystone ID
+            0xFE,
+            0x10, // Frame type
+            tx_power,
+            url_scheme.to_number(),
+        ];
+
+        let mut url_bytes = url.as_bytes();
```

Comment:
This matching should be done on a percent-encoded URL string to ensure that the string doesn't include any reserved bytes (e.g. `' '`). The [`url` create](https://crates.io/crates/url) might be suitable here.

Reply:
Do we even need the advertisement packets? It's not a part of the C API anyways...